### ISSUE TITLE
Can not use charts as dependencies

### ISSUE THREAD

**Issue by jmesnil** [openshift-helm-charts/charts#773]
Hi, I'm trying to create an application chart that depends the `postgresql-imagestreams`, `postgresql-persistent` and `eap74` chart.

My chart.yaml looks like:

```
apiVersion: v2
name: eap-with-postgresq
description: Build and deploy JBoss EAP 7.4 + PostgreSQL applications on OpenShift
type: application
version: 1.0.0

dependencies:
- name: redhat-eap74
  version: 1.1.1
  repository: https://charts.openshift.io
- name: redhat-postgresql-imagestreams
  version: 0.0.1
  repository: https://charts.openshift.io
- name: redhat-postgresql-persistent
  version: 0.0.2
  repository: https://charts.openshift.io
```

When I do a `helm dep up`, the dependencies are downloaded but their names do no match the dependencies names:

```
$ ls charts
eap74-1.1.1.tgz                         postgresql-imagestreams-0.0.1.tgz       postgresql-persistent-0.0.2.tgz
```

Then, when I package my application, it fails:

```
$ helm package .
Error: found in Chart.yaml, but missing in charts/ directory: redhat-eap74, redhat-postgresql-imagestreams, redhat-postgresql-persistent
```

Indeed, the name of the chart (eg `redhat-eap74` does not match the archive `eap74-1.1.1.tgz` and the chart inside the archive `eap74`).
I was not able to find a solution that work as the dependency name `redhat-eap74` will not match the actual chart name inside the archive `eap74`

I think this might related to a recent change as the previous release of `eap74` got the vendor prefixed:

```
$ curl https://charts.openshift.io/index.yaml | grep eap74
redhat-eap74:
    name: eap74
    - https://github.com/openshift-helm-charts/charts/releases/download/redhat-eap74-1.1.0/redhat-eap74-1.1.0.tgz
    name: eap74
    - https://github.com/openshift-helm-charts/charts/releases/download/redhat-eap74-1.1.1/eap74-1.1.1.tgz
```

but even previous version (1.1.0) would fail as the name of the chart  inside the archive (`eap74`) would not match the index name (`redhat-eap74`)

If I change the Helm repo for `eap74` to use our own that indexed it with `eap74` in the `index.yaml`, it works:

```
- name: eap74
  version: 1.1.0
  repository: https://jbossas.github.io/eap-charts/ 
```

**Comment by komish** [comment]
The infinispan chart has the same issue, as noted in the above duplicate issue ^

**Comment by komish** [comment]
Some additional information, I see that this was reported before https://bugzilla.redhat.com/show_bug.cgi?id=2081762. That issue is closed as WONTFIX, but a workaround was presented at the time. We'll do another evaluation on this subject as well, so leaving this open for now.

**Comment by jmesnil** [comment]
@komish thanks for the update.

The workaround is not applicable as it does not allow automation of Helm packaging.

**Comment by ryanemerson** [comment]
The workaround is also not applicable for Infinispan and we're getting an increasing number of users who are running into this issue.

**Comment by komish** [comment]
We're looking at the options. Unfortunately, this isn't a particularly simple lift. As we make traction, we'll keep this updated.

**Comment by komish** [comment]
Here's some initial thoughts based on my testing. Sharing here for the interested parties to follow along as we work through the logistics. This mostly aligns with what @jmesnil highlighted (thanks for the thorough bug report).

In my testing, it doesn't actually appear that the tarball or extracted Chart directory naming matters at all. The only thing that appears to matter is the Chart's `name` value as found in Chart.yaml, and the entry in the index for that same chart. 

- `helm dep up` checks your dependency listing and pulls from the `repository` value the chart with `name`. 
- `helm package ...` ensures the presence of the dependencies you specified in Chart.yaml exist in the `charts/` directory.

Modifying your Chart.yaml to match a dependency name to the actual name of the Chart as you have it pulled locally breaks `helm dep up` but fixes `helm package ...`.  This presents automation challenge @jmesnil points out (heard, loud and clear).

It's been stated (though potentially not here) that the index entry for these charts is organized in the way that it is to allow for charts to exist with the same name across our various contributors.

To that end, the simplest solution may be to enforce that the chart's name matches the index entry we intend to use for it. On the other hand, I acknowledge that this may not be ideal for Chart developers who publish various places from having to rename their chart to match our index policy.

The alternate path (checking for duplicate chart names, otherwise using the chart's name itself as the index entry for it) carries the potential for significant impact on running workloads.

We'll continue investigating. Keep an eye out here for updates. More to come as we have it.

**Comment by komish** [comment]
Hey folks,

Just wanted to throw in a quick update that we're actually considering migrating index entries to match the chart name vs. imposing a chart name standard (which we didn't love, but felt less impactful).

In my initial testing of migrating the chart entries, I actually saw minimal to no impact at all to running workloads. If that ends up being the case, then we'll likely go that route which will fix the dependency and also bring our index up to spec.

More to come as we have it, but just wanted to post a quick update to let interested parties know this is top of mind.

**Comment by komish** [comment]
Hey folks. Just another update on this that we're working through the development necessary to update the pipeline for changes related to this bugfix.

While we obviously have no duplicate entries in our index, and would continue not to have any after we migrate the index to use the proper format, we do have duplicate projects in our repository. That is, chart projects that have the same name, provided by different vendors. 

That said, these projects are just projects, and don't seem to contain any helm chart submissions. In a majority of cases, these projects are Red Hatters testing various things. We'll work to clean those things up and continue marching forward.

In summary, we're still working on getting this delivered as quickly as we can. Thanks for your patience.

**Comment by komish** [comment]
@pittar Yes, we're still working through some internal logistics.

**Comment by andcip** [comment]
Hi, 
just another report, same issue for jenkins chart.
More details:
I wanted to insert this chart in a GitOps repo:
```
helmCharts:
  - name: redhat-jenkins
    repo: https://charts.openshift.io
    version: "0.0.3"
    releaseName: jenkins
    namespace: devops
    valuesFile: values.yaml
    valuesInline:
      fullnameOverride: jenkins
```

running the command:
` oc kustomize --enable-helm .`
I found downloaded chart on dir jenkins, but the command fail with the error:

error: Error: path "/..../redhat-jenkins" not found

**Comment by komish** [comment]
Thanks for the report @andcip. I know it's taking some time, but we are still progressing here.

**Comment by komish** [comment]
Just another update that we're still working on this issue. We're waiting for some internal decisions to be made before we can continue move forward.

**Comment by komish** [comment]
Just an update that the work to resolve this continues. More to come as we have more information.

**Comment by komish** [comment]
Hey folks,

As of release [1.7.0](https://github.com/openshift-helm-charts/charts/commit/c5e684da798e2b1e564259b077293f656fe54071) this issue should be resolved (with certain caveats, noted below). You should now be able to depend on a partner charts from our index within your Chart.yaml, and be able to `helm dep up`, and `helm package` your chart. We also expect GitOps workflows to work as expected.

For a restatement (effectively similar information to what's captured in this issue) on _how_ we resolved this, please feel free to read our blog content [here](https://connect.redhat.com/en/blog/enabling-chart-dependency-within-red-hat-openshift-certified-helm-chart-index).

Finally, the caveat: There are some charts provided by Red Hat teams in this index that are still pending updates in order to be listed as a dependency. Some Red Hat charts were addressed ahead of the release, but others are just pending a new release version to land as "patched" in our index before they can be used as dependencies.

We appreciate your patience as we chose and applied an appropriate path forward for this issue.
